New page for queries

max-torch opened the pull request:
### In this PR

- A page navigation bar has been added to the header section
- A new page called Query, containing features from https://github.com/wesnoth/wesnoth-multiplayer-data-dashboard/tree/Table-and-Figures-Combined-early-build, has been created.
- On the Query page, when data is about to be fetched from the MariaDB database, if the number of data rows that would be fetched exceeds 5000, the data will not be fetched/processed and a popup will inform the user to reduce the scope of their query.
- Callbacks for UI components that are defined across three files: `layout.py`, `aggregate_statistics.py`, and `query.py` are all inside of `app.py` and have been separated from each other using comments. 

### Screenshots

Page Navigation Bar
![image](https://github.com/wesnoth/wesnoth-multiplayer-data-dashboard/assets/41433185/f7689f3d-d4bc-49a8-9591-844a83535be4)

Query Page
![image](https://github.com/wesnoth/wesnoth-multiplayer-data-dashboard/assets/41433185/d6ef6f47-46bd-4e8f-9ba5-b5c7477f0dc5)

Popup when the data to be fetched is too large
![image](https://github.com/wesnoth/wesnoth-multiplayer-data-dashboard/assets/41433185/96acb3fe-e5db-4fc6-b858-d36f554e9579)

max-torch committed:
Create query.py

max-torch committed:
Update app.py

Create callbacks for query.py

max-torch committed:
Add a navigation bar to the header

max-torch committed:
Limit query size for the table and histogram

max-torch committed:
Update query.py

Fix up appearance of imports

max-torch committed:
Update query.py

whitespace cleaning

max-torch committed:
Make the ids of elements across pages is different

Had to do this to solve an error where callbacks that where the app was attempting to trigger callbacks for other pages that were not the current page.
See https://community.plotly.com/t/dash-v1-11-0-release-introducing-pattern-matching-callbacks/37592/24

Pentarctagon commented:
What is the Query Page screenshot meant to show?

max-torch commented:
> What is the Query Page screenshot meant to show?

It is showing how the new page, called Query, looks like. In this page, there is a date range picker, a table whose data can be exported, a total games card, and a histogram.

Selecting a date range will trigger fetching data that will then be shown on the table and visualized in the histogram

max-torch committed:
Define browser page titles

soliton- commented on app.py:
```diff
@@ -496,13 +499,141 @@ def update_public_chart(start_date, end_date):
+    mariadb_connection = connect_to_mariadb()
+    cursor = mariadb_connection.cursor()
+    cursor.execute(
+        "SELECT * FROM tmp_game_info WHERE START_TIME BETWEEN ? AND ?", (start_date, end_date))
```
Table name is hard coded again.

soliton- commented on app.py:
```diff
@@ -496,13 +499,141 @@ def update_public_chart(start_date, end_date):
+        raise PreventUpdate
+    
+    # Inform the user that the query cannot be processed because the size of the data to process exceeds limitations.
+    if int(total_games.replace(",", "")) > 5000:
```
What is the replace for?

The limit should be configurable.

max-torch replied:
> What is the replace for?

I get an exception when applying `int()` to convert `total_games` from a string to an integer because of the existence of comma characters in the string when it represents numbers greater than 1,000. The `replace()` gets rid of any commas in the `total_games` before it is converted to an integer.

soliton- replied:
Is this using the string shown on the page? Why is this not using the actual number?

A random replace that works with one specific number formatting of some locale is not great.

max-torch replied:
Using the string. I didn't know that this wasn't great. I'll see if it is possible to use the actual number.

soliton- commented on layout.py:
```diff
@@ -63,35 +63,51 @@ def create_app_layout():
+                                id="nav-list",
+                                children=[
+                                    html.Li(dcc.Link("Statistics", href="/dashboard/")),
+                                    html.Li(dcc.Link("Query", href="/dashboard/query")),
```
Hard coded path instead of using `url_base_pathname`.

soliton- commented:
This PR can be tested on https://replays.wesnoth.org/dashboard-dev/query/.

soliton- commented:
Btw, I think design wise it'd look nicer with the games count on the top right next to the date range.

max-torch committed:
Update app.py

- Unhardcode a hardcoded reference to a database table name

max-torch committed:
Unhardcode the query row limit

max-torch committed:
Use integer of Total Games Value instead of a string for interaction between callbacks

Method used is documented here:
https://dash.plotly.com/sharing-data-between-callbacks#example-1---storing-data-in-the-browser-with-dcc.store

max-torch committed:
Update layout.py

Unhardcoded url base pathname

max-torch committed:
Put the Total Games card on the far right in the same row as the date range picker

max-torch commented:
Just added 5 commits that address all the latest feedback @soliton- 
Commit names (expand full commit name) accurately reflect commit contents.
Ready for another review.

soliton- commented:
The limit can continue to default to 5000. Just a config value overrides it. Otherwise looks good to me.

max-torch committed:
Make 5000 the default query row limit

max-torch commented:
Will merge to main in one day.

soliton- commented:
I would consolidate all the config handling in one function including the env and defaults instead of duplicating that.

max-torch committed:
Update footer_technology_stack.md

Add pandas since it's used now by the callbacks of the Query page.

max-torch committed:
Update user_guide.md

Less technical wording

max-torch committed:
Update user_guide.md

Restore sections that talk about the restored table and histogram

max-torch committed:
Update test_app.py

max-torch committed:
Update README.md

Mention that it's better for `port` in the config to be an integer instead of a string for tests to work properly.

max-torch committed:
Update test_app.py

Restore update tests for table and histogram

max-torch commented:
All tests are passing when run locally.
![image](https://github.com/wesnoth/wesnoth-multiplayer-data-dashboard/assets/41433185/d6a8368c-a3d2-4d0f-9ebc-23075b3fec1e)

max-torch commented:
> I would consolidate all the config handling in one function including the env and defaults instead of duplicating that.

Let's have a new issue #16  and PR for this

max-torch commented:
This PR closes #8
